scripts/src/java/org/oppia/android/scripts/ci/ComputeChangedFiles.kt

```diff
@@ -0,0 +1,315 @@
+
+/** Utility used to compute changed files. */
+class ComputeChangedFiles(
+  private val scriptBgDispatcher: ScriptBackgroundCoroutineDispatcher,
+  val maxFileCountPerLargeShard: Int = MAX_FILE_COUNT_PER_LARGE_SHARD,
+  val maxFileCountPerMediumShard: Int = MAX_FILE_COUNT_PER_MEDIUM_SHARD,
+  val maxFileCountPerSmallShard: Int = MAX_FILE_COUNT_PER_SMALL_SHARD,
+  val commandExecutor: CommandExecutor =
+    CommandExecutorImpl(
+      scriptBgDispatcher, processTimeout = 5, processTimeoutUnit = TimeUnit.MINUTES
+    )
+) {
+  private companion object {
+    private const val GENERIC_FILE_BUCKET_NAME = "generic"
+  }
+
+  /**
+   * Computes a list of files to run.
+   *
+   * @param pathToRoot the absolute path to the working root directory
+   * @param pathToOutputFile the absolute path to the file in which the encoded Base64 file bucket
+   *     protos should be printed
+   * @param baseCommit see [GitClient]
+   * @param computeAllFilesSetting whether all files should be outputted versus only the ones which
+   *     are changed by local changes in the repository
+   */
+  fun compute(
+    pathToRoot: String,
+    pathToOutputFile: String,
+    baseCommit: String,
+    computeAllFilesSetting: Boolean
+  ) {
+    val rootDirectory = File(pathToRoot).absoluteFile
+    check(rootDirectory.isDirectory) { "Expected '$pathToRoot' to be a directory" }
+    check(rootDirectory.list()?.contains("WORKSPACE") == true) {
+      "Expected script to be run from the workspace's root directory"
+    }
+
+    println("Running from directory root: $rootDirectory.")
+
+    val gitClient = GitClient(rootDirectory, baseCommit, commandExecutor)
+    println("Current branch: ${gitClient.currentBranch}.")
+    println("Most recent common commit: ${gitClient.branchMergeBase}.")
+
+    val currentBranch = gitClient.currentBranch.lowercase(Locale.US)
+    val changedFiles = if (computeAllFilesSetting || currentBranch == "develop") {
+      computeAllFiles(rootDirectory, pathToRoot)
+    } else computeChangedFilesForNonDevelopBranch(gitClient, rootDirectory)
+
+    val ktFiles = changedFiles.filter { it.endsWith(".kt") }
+    val filteredFiles = filterFiles(ktFiles)
+
+    val changedFileBuckets = bucketFiles(filteredFiles)
+    val encodedFileBucketEntries = changedFileBuckets
+      .associateBy { it.toCompressedBase64() }
+      .entries.shuffled()
+
+    File(pathToOutputFile).printWriter().use { writer ->
+      encodedFileBucketEntries.forEachIndexed { index, (encoded, bucket) ->
+        writer.println("${bucket.cacheBucketName}-shard$index;$encoded")
+      }
+    }
+  }
+
+  private fun computeAllFiles(
+    rootDirectory: File,
+    pathToRoot: String
+  ): List<String> {
+    val searchFiles = RepositoryFile.collectSearchFiles(
+      repoPath = pathToRoot,
+      expectedExtension = ".kt",
+    )
+
+    return searchFiles
+      .filter { it.extension == "kt" && !it.nameWithoutExtension.endsWith("Test") }
+      .map { it.toRelativeString(rootDirectory) }
```

Comment:
It seems worth it to me to also add tests for ensuring that changes to test and non-Kotlin files are excluded from results.

Reply:
Got it added the test case:
- testUtility_featureBranch_computeAllFiles_filtersNonKotlinFiles